### PR TITLE
[spoiler] feat: adding Spoiler plugin for image, gallery and video

### PR DESCRIPTION
**Pull request by barboaron**

### CODE REVIEW

File: packages/common/web/statics/locale/messages_en.json [hunk 1]
```diff
@@ -5,6 +5,12 @@
   "TextHighlightButton_Tooltip": "Highlight color",
   "ColorPickerButtonLabel_Cancel": "Cancel",
   "ColorPickerButtonLabel_Update": "Save",
+  "Spoiler_Reveal_Image_Placeholder" : "This image contains a spoiler",
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/common/web/statics/locale/messages_en.json]: removed

File: packages/editor-common/web/src/Base/buttons/BlockSpoilerButton.jsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import InlineToolbarButton from '../../Components/InlineToolbarButton';
+import { SpoilerButtonIcon } from '../../Icons';
+
+class BlockSpoilerButton extends Component {
+  handleClick = event => {
+    event.preventDefault();
+    this.wrapBlockInSpoiler();
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor-common/web/src/Base/buttons/BlockSpoilerButton.jsx]: done

File: packages/editor-common/web/src/Base/buttons/BlockSpoilerButton.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  wrapBlockInSpoiler = () => {
+    const { pubsub } = this.props;
+    const { spoiler = {} } = this.props.pubsub.get('componentData')?.config;
+
+    pubsub.setBlockData({
+      key: 'componentSpoiler',
+      item: { ...spoiler, enabled: this.isActive ? null : true },
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor-common/web/src/Base/buttons/BlockSpoilerButton.jsx]: done

File: packages/editor/web/src/RichContentEditor/Decorators/spoilerWrapper.jsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+export default (BlockSpoilerComponent, config) => WrappedComponent => {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor/web/src/RichContentEditor/Decorators/spoilerWrapper.jsx:4]: done

File: packages/editor/web/src/RichContentEditor/Decorators/spoilerWrapper.jsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      dataHook: PropTypes.string,
+      onClick: PropTypes.func,
+      className: PropTypes.string,
+      setFocusToBlock: PropTypes.func,
+      setInPluginEditingMode: PropTypes.func,
+      store: PropTypes.object,
+      block: PropTypes.object,
+      size: PropTypes.object,
+    };
+    static defaultProps = {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor/web/src/RichContentEditor/Decorators/spoilerWrapper.jsx:22]: removed

File: packages/editor/web/src/RichContentEditor/createPlugins.jsx [hunk 6]
```diff
@@ -28,14 +30,29 @@ const createPlugins = ({ plugins, context, commonPubsub }) => {
 
   const pluginDefaults = {};
 
+  const spoilerPlugin = plugins?.filter(plugin => plugin.BlockSpoilerComponent)?.[0];
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor/web/src/RichContentEditor/createPlugins.jsx:33]: done

File: packages/plugin-gallery/web/src/gallery-component.jsx [hunk 7]
```diff
@@ -33,14 +33,17 @@ class GalleryComponent extends PureComponent {
 
   componentWillReceiveProps(nextProps) {
     const { componentData, componentState } = this.props;
+    const spoiler_description = componentData?.config?.spoiler?.description;
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-gallery/web/src/gallery-component.jsx:36]: we don't need this with the new decorator (wrapper), removed

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    const { styles, hasSpoiler, isReveal } = this.state;
+
+    let className = '';
+    if (hasSpoiler && !isReveal) {
+      className = pluginType === 'Gallery' ? styles.hideBlock_gallery : styles.hideBlock;
+    }
+
+    return (
+      <div
+        ref={ref => (this.element = ref)}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:108]: yes, in 'componentDidMount'

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        data-hook={dataHook}
+        className={classNames(styles.spoilerWrapper, this.props.className)}
+        style={size}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:111]: yes, we need this for the viewer

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+      disabledRevealSpoilerBtn,
+      setFocusToBlock,
+      EditableSpoilerDescription,
+      componentData,
+      pluginType,
+      block,
+      blockProps,
+      store,
+      setInPluginEditingMode,
+      size,
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:66]: fixed

File: packages/plugin-spoiler/web/src/RevealButton.jsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    const oldValue = this.props.value;
+    this.props.onChange?.(e.target.value);
+    if (e.target.scrollWidth > 228 && oldValue.length < e.target.value.length) {
+      this.props.onChange?.(oldValue);
+    }
+  };
+
+  render() {
+    const {
+      EditableSpoilerDescription,
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/RevealButton.jsx:31]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (shaulgo) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:104]: yup, but seem like it shouldn't be used here. If I see correctly it's used to get the size. You can do that where it's needed.

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+import SpoilerContainer from './SpoilerContainer';
+import classNames from 'classnames';
+import styles from '../statics/styles/spoiler.scss';
+
+class BlockSpoilerComponent extends React.Component {
+  constructor(props) {
+    super(props);
+    const { theme } = props;
+    this.state = {
+      hasSpoiler: props.componentData?.config?.spoiler?.enabled || false,
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+          role="button"
+          tabIndex={0}
+          onKeyDown={this.onKeyDown}
+        >
+          {children}
+          {!isReveal && (
+            <div
+              role="none"
+              className={pluginType === 'Gallery' ? styles.overlay_gallery : styles.overlay}
+            />
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:116]: <Overlay hideOverlay={isReveal}/>

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 14]
```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { mergeStyles, GlobalContext } from 'wix-rich-content-common';
+import SpoilerContainer from './SpoilerContainer';
+import classNames from 'classnames';
+import styles from '../statics/styles/spoiler.scss';
+
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:7]: const Overlay = {hide} => hide ? ( <div role="none" className={pluginType === 'Gallery' ? styles.overlay_gallery : styles.overlay}/> : null;

File: packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    element.style.height = 'auto';
+    element.style.height = element.scrollHeight + 'px';
+    this.textAreaElement = element;
+  }
+
+  render() {
+    const { className, disabled, ...otherProps } = this.props;
+    const InputComponent = (
+      <textarea
+        rows="1"
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx:19]: dir="auto"

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+export default config => WrappedComponent => {
+  return class spoilerWrapper extends Component {
+    constructor(props) {
+      super(props);
+      this.state = {
+        hasSpoiler: props.componentData?.config?.spoiler?.enabled || false,
+      };
+    }
+
+    static WrappedComponent = WrappedComponent;
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:14]:
```suggestion
```

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+      setFocusToBlock: PropTypes.func,
+      setInPluginEditingMode: PropTypes.func,
+      store: PropTypes.object,
+      block: PropTypes.object,
+    };
+
+    componentWillReceiveProps(props) {
+      const hasSpoiler = props.componentData?.config?.spoiler?.enabled || false;
+      this.setState({ hasSpoiler });
+    }
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:33]:
```suggestion
```

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+      const pluginType = type[0].toUpperCase() + type.slice(1);
+      const {
+        componentData,
+        theme,
+        onClick,
+        className,
+        setInPluginEditingMode,
+        blockProps,
+      } = this.props;
+      const { hasSpoiler } = this.state;
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:65]:
```suggestion
      const hasSpoiler = props.componentData?.config?.spoiler?.enabled || false;
```

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+import PropTypes from 'prop-types';
+import BlockSpoilerComponent from './BlockSpoilerComponent';
+
+export default config => WrappedComponent => {
+  return class spoilerWrapper extends Component {
+    constructor(props) {
+      super(props);
+      this.state = {
+        hasSpoiler: props.componentData?.config?.spoiler?.enabled || false,
+      };
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:11]:
```suggestion
```

File: packages/plugin-spoiler/web/src/EditableTextWrapper.jsx [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+  handleKeyPress = e => {
+    const { setFocusToBlock, value } = this.props;
+    if (e.key === 'Enter' && setFocusToBlock && value !== '') {
+      this.handleBlur();
+      setFocusToBlock();
+    }
+  };
+
+  onChange = e => this.props.onChange?.(e);
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/EditableTextWrapper.jsx:34]:
```suggestion
```

File: packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+export default class SpoilerDescriptionInput extends Component {
+  componentWillReceiveProps() {
+    // eslint-disable-next-line react/no-find-dom-node
+    const element = this.textAreaElement || findDOMNode(this);
+    element.style.height = 'auto';
+    element.style.height = element.scrollHeight + 'px';
+    this.textAreaElement = element;
+  }
+
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx:14]: onChange = e => props.onChange(e)

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+      const { hasSpoiler } = this.state;
+
+      return hasSpoiler ? (
+        <BlockSpoilerComponent
+          disabledRevealSpoilerBtn
+          isEditableText
+          handleButtonContentChange={this.handleButtonContentChange}
+          handleDescriptionChange={this.handleDescriptionChange}
+          pluginType={pluginType}
+          componentData={componentData}
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:74]: transform comopnentData to the values you need

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+      setInPluginEditingMode,
+      handleButtonContentChange,
+      handleDescriptionChange,
+    } = this.props;
+
+    return (
+      !isReveal && (
+        <SpoilerContainer
+          styles={styles}
+          width={width || this.props.width}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:66]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+      disabledRevealSpoilerBtn,
+      setFocusToBlock,
+      isEditableText,
+      componentData,
+      pluginType,
+      setInPluginEditingMode,
+      handleButtonContentChange,
+      handleDescriptionChange,
+    } = this.props;
+
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:61]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+      )
+    );
+  };
+
+  onKeyDown = e => {
+    if (e.key === 'Enter') {
+      this.handleClick(e);
+    }
+  };
+
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:87]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+    this.setState({ height, width });
+  }
+
+  componentDidUpdate() {
+    const { offsetWidth: width, offsetHeight: height } = this?.element;
+    if (this.state.height !== height || this.state.width !== width) {
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState({ height, width });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:30]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+    const { children, pluginType, dataHook, width } = this.props;
+    const { styles, isReveal } = this.state;
+    let className = '';
+    if (!isReveal) {
+      className = pluginType === 'Gallery' ? styles.hideBlock_gallery : styles.hideBlock;
+    }
+
+    return (
+      <div
+        ref={ref => (this.element = ref)}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:98]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+    return (
+      !isReveal && (
+        <SpoilerContainer
+          styles={styles}
+          width={width || this.props.width}
+          height={height}
+          setInPluginEditingMode={setInPluginEditingMode}
+          setFocusToBlock={setFocusToBlock}
+          pluginType={pluginType}
+          isEditableText={isEditableText}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:71]: done

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+          disabledRevealSpoilerBtn={disabledRevealSpoilerBtn}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:72]: done

File: packages/plugin-spoiler/web/src/RevealButton.jsx [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+      isEditableText,
+      isMobile,
+      className,
+      disabledRevealSpoilerBtn,
+      onRevealSpoiler,
+      setInPluginEditingMode,
+      setFocusToBlock,
+      value,
+    } = this.props;
+    const dataHook = !disabledRevealSpoilerBtn && 'revealSpoilerBtn';
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/RevealButton.jsx:25]: done

File: packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx (hunk 15 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx:19]: done

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx (hunk 16 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:14]: removed

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx (hunk 17 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:33]: removed

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx (hunk 18 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:65]: done

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx (hunk 19 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:11]: removed

File: packages/plugin-spoiler/web/src/EditableTextWrapper.jsx (hunk 20 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/EditableTextWrapper.jsx:34]: removed

File: packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx (hunk 21 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/SpoilerDescriptionInput.jsx:14]: added

File: packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx (hunk 22 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerEditorWrapper.jsx:74]: I pass the componentData to BlockSpoilerComponent from the editor wrapper and the viewer wrapper, and then I pass to it's children only the values I need, like we talked yesterday, it's ok?

File: packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx (hunk 14 above)
Comment (barboaron) [reply packages/plugin-spoiler/web/src/BlockSpoilerComponent.jsx:7]: done

**Pull request closed without merging**